src/PureScript/Backend/Optimizer/Tracer/Printer.purs

```diff
@@ -0,0 +1,516 @@
+    let
+      printArg arg =
+        D.words
+          [ D.text "\\" <> uncurry printLocal arg
+          , D.text "->" <> D.space
+          ]
+    printIndentedParens' (foldMap1 printArg args) $ printBackendExpr body
+  UncurriedApp fn args ->
+    wrapInParens $ D.words $ map printBackendExpr $ Array.cons fn args
+  UncurriedAbs args body ->
+    printIndentedParens'
+      (D.text "\\" <> (D.words $ map (uncurry printLocal) args) <> D.space <> D.text "->")
+      $ printBackendExpr body
+  UncurriedEffectApp fn args ->
+    D.text "(#" <> (D.words $ map printBackendExpr $ Array.cons fn args) <> D.text "#)"
+  UncurriedEffectAbs args body ->
+    D.lines
+      [ D.text "(#\\" <> (D.words $ map (uncurry printLocal) args) <> D.space <> D.text "->"
+      , D.indent $ printBackendExpr body
+      , D.text "#)"
+      ]
+  Accessor expr accessor ->
+    printIndentedParens (printBackendExpr expr) <> printBackendAccessor accessor
+  Update expr propArr -> do
+    let sep = D.break <> D.text ", "
+    printIndentedParens $
+      D.lines
+        [ printBackendExpr expr
+        , D.indent $ D.lines
+            [ D.text "{"
+            , D.foldWithSeparator sep $ map (printProp printBackendExpr) propArr
+            , D.text "}"
+            ]
+        ]
+
+  CtorSaturated qi _ ctorName _ values ->
+    wrapInParens $
+      D.words
+        [ printQualifiedIdent qi <> D.text "." <> printProperName ctorName
+        , D.words $ map (printBackendExpr <<< snd) values
+        ]
+
+  CtorDef _ _ ctorName [] ->
+    printLet "let" (printIdent ctorName)
+      $ printRecord
+          [ (Prop "tag" $ ExprSyntax mempty $ Lit $ LitString $ (\(Ident i) -> i) ctorName) ]
+
+  CtorDef _ _ ctorName args ->
+    printLet "let" (printIdent ctorName)
+      $ printIndentedParens'
+          (guard (not $ Array.null args) $ D.words [ D.text "\\", D.words $ map D.text args, D.text "->" ])
+      $ printRecord
+      $ Array.cons (Prop "tag" $ ExprSyntax mempty $ Lit $ LitString $ (\(Ident i) -> i) ctorName)
+      $ args <#> \arg -> Prop arg $ ExprSyntax mempty $ Var $ Qualified Nothing $ Ident arg
+
+  LetRec lvl bindings body ->
+    D.lines
+      [ D.text "letrec"
+      , D.indent
+          $ D.lines
+          $ bindings <#>
+              ( \(Tuple ident binding) ->
+                  printLet "let" (printLocal (Just ident) lvl) $ printBackendExpr binding
+              )
+      , printBackendExpr body
+      ]
+
+  Let mbIdent lvl binding body ->
+    D.lines
+      [ printLet "let" (printLocal mbIdent lvl) $ printBackendExpr binding
+      , printBackendExpr body
+      ]
+
+  EffectBind mbIdent lvl binding body ->
+    D.lines
+      [ printLet "letEffect" (printLocal mbIdent lvl) $ printBackendExpr binding
+      , printBackendExpr body
+      ]
+  EffectPure a ->
+    wrapInParens $ D.words [ D.text "effectPure", printBackendExpr a ]
+
+  EffectDefer a ->
+    wrapInParens $ D.words [ D.text "effectDefer", printBackendExpr a ]
```

Comment:
I wasn't sure how I should handle these two cases, so I acted like they were a special primitive function.